core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -510,6 +512,21 @@ class Partition(val topic: String,
     }
   }
 
+  /**
+    * @param leaderEpoch Requested leader epoch
+    * @return The last offset of messages published under this leader epoch.
+    */
+  def lastOffsetForLeaderEpoch(leaderEpoch: Int): EpochEndOffset = {
+    inReadLock(leaderIsrUpdateLock) {
+      leaderReplicaIfLocal match {
+        case Some(leaderReplica) =>
+          new EpochEndOffset(NONE, leaderReplica.epochs.get.endOffsetFor(leaderEpoch))
+        case None =>
+          new EpochEndOffset(NOT_LEADER_FOR_PARTITION, UNDEFINED_EPOCH_OFFSET)
```

Comment:
Nit: `map` followed by `getOrElse` is a little more concise, but OK like this if you prefer. Same applies for `lastOffsetForLeaderEpoch`.